### PR TITLE
edgehub: Unit tests for edgehub/controller.go

### PR DESCRIPTION
**Pull request by sujithsimon22**
1. Unit tests for edgehub/controller.go
2. Added mockfile for edgehub/clients/adapter.go

### CODE REVIEW

**Comment by sids-b** [comment]
@sujithsimon22 ,we have a PR fixing all the golint warnings, can you confirm once if there aren't any ?

**Comment by sujithsimon22** [comment]
> @sujithsimon22 ,we have a PR fixing all the golint warnings, can you confirm once if there aren't any ?

@sids-b I have checked, there are no golint warnings in the files that have been added in this PR :)

File: pkg/edgehub/controller_test.go [hunk 1]
```diff
@@ -0,0 +1,869 @@
+package edgehub
```
Comment (m1093782566) [inline pkg/edgehub/controller_test.go]: KubeEdge License head please.

File: pkg/edgehub/controller_test.go [hunk 2]
```diff
@@ -0,0 +1,885 @@
...
+			Organization: []string{"testor"},
+		},
+		NotBefore:   time.Now(),
+		NotAfter:    time.Now().AddDate(5, 5, 5),
+		ExtKeyUsage: []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth, x509.ExtKeyUsageServerAuth},
+		KeyUsage:    x509.KeyUsageDigitalSignature | x509.KeyUsageCertSign,
+	}
+
+	// generate private key
+	privatekey, err := rsa.GenerateKey(rand.Reader, 2048)
```
Comment (sids-b) [inline pkg/edgehub/controller_test.go]: camel case please

File: pkg/edgehub/controller_test.go [hunk 3]
```diff
@@ -0,0 +1,885 @@
...
+
```
Comment (sids-b) [inline pkg/edgehub/controller_test.go]: remove blank line

File: pkg/edgehub/controller_test.go [hunk 4]
```diff
@@ -0,0 +1,885 @@
...
+	if err != nil {
+		fmt.Println(err)
+	}
+
```
Comment (sids-b) [inline pkg/edgehub/controller_test.go]: remove blank line

File: pkg/edgehub/controller_test.go [hunk 5]
```diff
@@ -0,0 +1,885 @@
...
+	publickey := &privatekey.PublicKey
```
Comment (sids-b) [inline pkg/edgehub/controller_test.go]: camel case please

File: pkg/edgehub/controller_test.go [hunk 6]
```diff
@@ -0,0 +1,885 @@
...
+		fmt.Println(err)
+	}
+	err = os.MkdirAll("/tmp/kubeedge/testData", 0777)
+	if err != nil {
+		fmt.Println(err)
+	}
+	pkey := x509.MarshalPKCS1PrivateKey(privatekey)
+
+	// this will create plain text PEM file.
+	certfile, _ := os.Create("/tmp/kubeedge/testData/edge.crt")
```
Comment (sids-b) [inline pkg/edgehub/controller_test.go]: handle error

File: pkg/edgehub/controller_test.go [hunk 7]
```diff
@@ -0,0 +1,885 @@
...
+	var pemkey = &pem.Block{
+		Type:  "CERTIFICATE",
+		Bytes: cert}
+	pem.Encode(certfile, pemkey)
+	certfile.Close()
+
+	// this will create plain text PEM file.
+	keyfile, _ := os.Create("/tmp/kubeedge/testData/edge.key")
```
Comment (sids-b) [inline pkg/edgehub/controller_test.go]: handle error

File: pkg/edgehub/controller_test.go [hunk 8]
```diff
@@ -0,0 +1,885 @@
...
+
+//TestInitial() tests the procurement of the cloudhub client
+func TestInitial(t *testing.T) {
+	tests := []struct {
+		name             string
+		fields           EdgeHubController
+		webSocketConfig  config.WebSocketConfig
+		controllerConfig config.ControllerConfig
+		ctx              *context.Context
+		wantErr          bool
```
Comment (sids-b) [inline pkg/edgehub/controller_test.go]: wantErr -> wantBool ??

File: pkg/edgehub/controller_test.go [hunk 9]
```diff
@@ -0,0 +1,885 @@
...
+}
+
+//TestSendToKeepChannel() tests the reception of response in the syncKeep channel
+func TestSendToKeepChannel(t *testing.T) {
+	tests := []struct {
+		name              string
+		fields            EdgeHubController
+		message           *model.Message
+		wantSyncKeeperErr bool
+		wantErr           bool
```
Comment (sids-b) [inline pkg/edgehub/controller_test.go]: wantBool ?? Applicable to all cases below

File: pkg/edgehub/controller_test.go [hunk 10]
```diff
@@ -0,0 +1,885 @@
...
+	}
+}
+
+//TestRouteToEdge() is used to test whether the message received from websocket is dispatched to the required modules
+func TestRouteToEdge(t *testing.T) {
+	initMocks(t)
+	tests := []struct {
+		name       string
+		fields     EdgeHubController
+		receiveErr bool
```
Comment (sids-b) [inline pkg/edgehub/controller_test.go]: receiveErr -> receiveBool ??

File: pkg/edgehub/controller_test.go [hunk 11]
```diff
@@ -0,0 +1,885 @@
...
+				stopChan := <-ehc.stopChan
+				if stopChan != struct{}{} {
+					t.Errorf("TestRouteToEdge error got: %v want: %v", stopChan, struct{}{})
+				}
+			} else {
+				mockAdapter.EXPECT().Receive().Return(*model.NewMessage("test").BuildRouter(ModuleNameEdgeHub, core.EdgedGroup, "", ""), nil).Times(1)
+				mockAdapter.EXPECT().Receive().Return(*model.NewMessage("test").BuildRouter(ModuleNameEdgeHub, core.TwinGroup, "", ""), nil).Times(1)
+				mockAdapter.EXPECT().Receive().Return(*model.NewMessage(""), errors.New("Connection Refused")).Times(1)
+				go ehc.routeToEdge()
+				stopChan := <-ehc.stopChan
```
Comment (sids-b) [inline pkg/edgehub/controller_test.go]: can we rename stopChan ? as it is not a channel

File: pkg/edgehub/controller_test.go [hunk 12]
```diff
@@ -0,0 +1,885 @@
...
+func TestSendToCloud(t *testing.T) {
+	initMocks(t)
+	msg := model.NewMessage("").BuildHeader("test_id", "", 1)
+	msg.Header.Sync = true
+	tests := []struct {
+		name    string
+		fields  EdgeHubController
+		message model.Message
+		wantErr bool
+		waitErr bool
```
Comment (sids-b) [inline pkg/edgehub/controller_test.go]: waitErr -> wait ?

File: pkg/edgehub/controller_test.go (hunk 8 above)
Comment (sujithsimon22) [reply pkg/edgehub/controller_test.go]: Here wantErr variable is used to signify whether it is a positive test case or not. It basically is used to tell the user whether the test case is expecting an error or not. If wantErr is true, it means that an error is expected in the test case(negative test case), if it is false then it signifies that no error is expected. I feel that wantBool might not convey the message properly and might not be easy to understand.

File: pkg/edgehub/controller_test.go (hunk 9 above)
Comment (sujithsimon22) [reply pkg/edgehub/controller_test.go]: Here wantErr variable is used to signify whether it is a positive test case or not. It basically is used to tell the user whether the test case is expecting an error or not. If wantErr is true, it means that an error is expected in the test case(negative test case), if it is false then it signifies that no error is expected. I feel that wantBool might not convey the message properly and might not be easy to understand.

File: pkg/edgehub/controller_test.go (hunk 10 above)
Comment (sujithsimon22) [reply pkg/edgehub/controller_test.go]: Here receiveErr variable is used to signify whether it is an error in receiving the message from the cloud. It basically is used to tell the user whether the test case is expecting an error or not. If receiveErr is true, it means that an error is expected in the test case, if it is false then it signifies that no error is expected in the reception of the message. I feel that receiveBool might not convey the message properly and might not be easy to understand.

File: pkg/edgehub/controller_test.go [hunk 13]
```diff
@@ -0,0 +1,889 @@
...
+				flag = false
+			} else {
+				w.Write([]byte("ws://127.0.0.1:20000"))
+			}
+		}
+	}))
+}
+
+//generateTestCertificate() generates fake certificates and stores them at a temporary location (/tmp/kubeedge/testData)
+func generateTestCertificate() {
```
Comment (rohitsardesai83) [inline pkg/edgehub/controller_test.go]: this function can be moved to a util file. the test code can simply make utility func calls and thus remain more cleaner.

File: pkg/edgehub/controller_test.go [hunk 14]
```diff
@@ -0,0 +1,889 @@
...
+	}
+	publicKey := &privateKey.PublicKey
+
+	// create a self-signed certificate. template = parent
+	var parent = template
+	cert, err := x509.CreateCertificate(rand.Reader, template, parent, publicKey, privateKey)
+	if err != nil {
+		fmt.Println(err)
+	}
+	err = os.MkdirAll("/tmp/kubeedge/testData", 0777)
```
[the remark this answers is not in the document]
Comment (rohitsardesai83) [reply pkg/edgehub/controller_test.go]: this comment applies to all such occurrences in this test code.

File: pkg/edgehub/controller_test.go [hunk 15]
```diff
@@ -0,0 +1,889 @@
...
+func TestNewEdgeHubController(t *testing.T) {
+	tests := []struct {
+		name             string
+		want             *EdgeHubController
+		controllerConfig config.ControllerConfig
+	}{
+		{"TestNewEdgeHubController: Positive Test Case",
+			&EdgeHubController{
+				config: &config.ControllerConfig{
+					HeartbeatPeroid: 150 * time.Second,
```
Comment (rohitsardesai83) [inline pkg/edgehub/controller_test.go]: nit: HeartbeatPeroid -> HeartbeatPeriod.  code change is needed .

File: pkg/edgehub/controller_test.go [hunk 16]
```diff
@@ -0,0 +1,889 @@
...
+}
+
+//TestNewEdgeHubController() tests whether the EdgeHubController returned is correct or not
+func TestNewEdgeHubController(t *testing.T) {
+	tests := []struct {
+		name             string
+		want             *EdgeHubController
+		controllerConfig config.ControllerConfig
+	}{
+		{"TestNewEdgeHubController: Positive Test Case",
```
Comment (rohitsardesai83) [inline pkg/edgehub/controller_test.go]: can the name be more descriptive ? what is the positive test case ? any negative tests also we can write ?

File: pkg/edgehub/controller_test.go [hunk 17]
```diff
@@ -0,0 +1,889 @@
...
+			&EdgeHubController{
+				config: &config.ControllerConfig{
+					HeartbeatPeroid: 150 * time.Second,
+					RefreshInterval: 15 * time.Minute,
```
Comment (rohitsardesai83) [inline pkg/edgehub/controller_test.go]: if controller config's are reused , then better have constants for all config parameters or create a config object once and reuse.

File: pkg/edgehub/controller_test.go [hunk 18]
```diff
@@ -0,0 +1,889 @@
...
+			edgeHubConfig.CtrConfig = tt.controllerConfig
+			ehc := &EdgeHubController{
+				context:    tt.fields.context,
+				chClient:   tt.fields.chClient,
+				config:     tt.fields.config,
+				stopChan:   tt.fields.stopChan,
+				syncKeeper: tt.fields.syncKeeper,
+				keeperLock: tt.fields.keeperLock,
+			}
+			if err := ehc.initial(tt.ctx); (err != nil) != tt.wantErr {
```
Comment (sids-b) [inline pkg/edgehub/controller_test.go]: instead of just checking if err is nil or not, can we not assert the error ? This will make sure that the failure scenario for which the test case is designed for is achieved. 
Applicable to similar cases below

File: pkg/edgehub/controller_test.go [hunk 19]
```diff
@@ -0,0 +1,889 @@
...
+		NotBefore:   time.Now(),
+		NotAfter:    time.Now().AddDate(5, 5, 5),
+		ExtKeyUsage: []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth, x509.ExtKeyUsageServerAuth},
+		KeyUsage:    x509.KeyUsageDigitalSignature | x509.KeyUsageCertSign,
+	}
+
+	// generate private key
+	privateKey, err := rsa.GenerateKey(rand.Reader, 2048)
+	if err != nil {
+		fmt.Println(err)
```
Comment (sids-b) [inline pkg/edgehub/controller_test.go]: Let this function return error to make sure creation of certificates was success or not rather than just printing error.

File: pkg/edgehub/controller_test.go [hunk 20]
```diff
@@ -0,0 +1,889 @@
...
+			ehc := &EdgeHubController{
+				context:    tt.fields.context,
+				chClient:   tt.fields.chClient,
+				config:     tt.fields.config,
+				stopChan:   tt.fields.stopChan,
+				syncKeeper: tt.fields.syncKeeper,
+				keeperLock: tt.fields.keeperLock,
+			}
+			if tt.wantErr {
+				mockAdapter.EXPECT().Send(gomock.Any()).Return(errors.New("Connection Refused")).Times(1)
```
Comment (sids-b) [inline pkg/edgehub/controller_test.go]: instead of hardcoding this, can we do something we have done in this file
https://github.com/kubeedge/kubeedge/blob/master/pkg/devicetwin/dtclient/device_db_test.go#L380
This might help in avoiding if-else conditions in the code ?

**Comment by sids-b** [comment]
@sujithsimon22 , please update you fork to latest

File: pkg/edgehub/controller_test.go [hunk 21]
```diff
@@ -0,0 +1,820 @@
...
+
+// mockAdapter is a mocked adapter implementation
+var mockAdapter *edgehub.MockAdapter
+
+//init() starts the test server and generates test certificates for testing
+func init() {
+	newTestServer()
+	err := util.GenerateTestCertificate("/tmp/kubeedge/Certs/", "edge", "edge")
+	if err != nil {
+		fmt.Println("Error in creating fake certificates: ", err)
```
Comment (sids-b) [inline pkg/edgehub/controller_test.go]: panic instead of print

File: pkg/common/util/config.go [hunk 22]
```diff
@@ -78,3 +85,67 @@ func GenerateTestYaml(test interface{}, path, filename string) error {
...
+			Organization: []string{"testor"},
+		},
+		NotBefore:   time.Now(),
+		NotAfter:    time.Now().AddDate(5, 5, 5),
+		ExtKeyUsage: []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth, x509.ExtKeyUsageServerAuth},
+		KeyUsage:    x509.KeyUsageDigitalSignature | x509.KeyUsageCertSign,
+	}
+
+	// generate private key
+	privatekey, err := rsa.GenerateKey(rand.Reader, 2048)
```
Comment (sids-b) [inline pkg/common/util/config.go]: camel case please

File: pkg/common/util/config.go [hunk 23]
```diff
@@ -78,3 +85,67 @@ func GenerateTestYaml(test interface{}, path, filename string) error {
...
+	if err != nil {
+		return err
+	}
+	publickey := &privatekey.PublicKey
```
Comment (sids-b) [inline pkg/common/util/config.go]: camel case please

File: pkg/common/util/config.go [hunk 24]
```diff
@@ -78,3 +85,67 @@ func GenerateTestYaml(test interface{}, path, filename string) error {
...
+		return err
+	}
+	var pemkey = &pem.Block{
+		Type:  "CERTIFICATE",
+		Bytes: cert}
+	pem.Encode(certfile, pemkey)
+	certfile.Close()
+
+	// this will create plain text PEM file.
+	keyfile, err := os.Create(path + keyFileName + ".key")
```
Comment (sids-b) [inline pkg/common/util/config.go]: camel case please

File: pkg/common/util/config.go [hunk 25]
```diff
@@ -78,3 +85,67 @@ func GenerateTestYaml(test interface{}, path, filename string) error {
...
+		return err
+	}
+	err = os.MkdirAll(path, 0777)
+	if err != nil {
+		return err
+	}
+	pkey := x509.MarshalPKCS1PrivateKey(privatekey)
+
+	// this will create plain text PEM file.
+	certfile, err := os.Create(path + certFileName + ".crt")
```
Comment (sids-b) [inline pkg/common/util/config.go]: camel case please

File: pkg/common/util/config.go [hunk 26]
```diff
@@ -78,3 +85,67 @@ func GenerateTestYaml(test interface{}, path, filename string) error {
...
+		Bytes: cert}
+	pem.Encode(certfile, pemkey)
+	certfile.Close()
+
+	// this will create plain text PEM file.
+	keyfile, err := os.Create(path + keyFileName + ".key")
+	if err != nil {
+		return err
+	}
+	var pemkey1 = &pem.Block{
```
Comment (sids-b) [inline pkg/common/util/config.go]: camel case please. Can we rename pemkey1 to something else ?

File: pkg/edgehub/controller_test.go [hunk 27]
```diff
@@ -0,0 +1,820 @@
...
+			edgeHubConfig.CtrConfig = tt.controllerConfig
+			ehc := &Controller{
+				context:    tt.fields.context,
+				chClient:   tt.fields.chClient,
+				config:     tt.fields.config,
+				stopChan:   tt.fields.stopChan,
+				syncKeeper: tt.fields.syncKeeper,
+				keeperLock: tt.fields.keeperLock,
+			}
+			if err := ehc.initial(tt.ctx); err != tt.expectedError {
```
Comment (sids-b) [inline pkg/edgehub/controller_test.go]: Is it possible to add a negative test case here ?

File: pkg/common/util/config.go [hunk 28]
```diff
@@ -78,3 +85,67 @@ func GenerateTestYaml(test interface{}, path, filename string) error {
...
+		Subject: pkix.Name{
+			Country:      []string{"test"},
+			Organization: []string{"testor"},
+		},
+		NotBefore:   time.Now(),
+		NotAfter:    time.Now().AddDate(5, 5, 5),
+		ExtKeyUsage: []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth, x509.ExtKeyUsageServerAuth},
+		KeyUsage:    x509.KeyUsageDigitalSignature | x509.KeyUsageCertSign,
+	}
+
```
Comment (rohitsardesai83) [inline pkg/common/util/config.go]: extra new line ?

File: pkg/common/util/config.go (hunk 25 above)
Comment (rohitsardesai83) [inline pkg/common/util/config.go]: the two code blocks look similar , better we make a common function.

File: pkg/edgehub/controller_test.go [hunk 29]
```diff
@@ -0,0 +1,820 @@
...
+	"github.com/kubeedge/kubeedge/beehive/pkg/core/context"
+	"github.com/kubeedge/kubeedge/beehive/pkg/core/model"
+	"github.com/kubeedge/kubeedge/mocks/edgehub"
+	"github.com/kubeedge/kubeedge/pkg/common/util"
+	_ "github.com/kubeedge/kubeedge/pkg/devicetwin"
+	"github.com/kubeedge/kubeedge/pkg/edgehub/config"
+)
+
+const (
+	CertFile = "/tmp/kubeedge/Certs/edge.crt"
```
Comment (rohitsardesai83) [inline pkg/edgehub/controller_test.go]: Certs -> certs

File: pkg/edgehub/controller_test.go (hunk 27 above)
Comment (sujithsimon22) [reply pkg/edgehub/controller_test.go]: @sids-b The source code for initial() has been written in such a way that it is not possible to add a negative test case here

**Comment by sids-b** [comment]
rerunning the workflow...

File: pkg/edgehub/controller_test.go [hunk 30]
```diff
@@ -0,0 +1,820 @@
...
+		want             *Controller
+		controllerConfig config.ControllerConfig
+	}{
+		{"TestNewEdgeHubController: Testing if EdgeHubController is returned with correct values",
+			&Controller{
+				config: &config.ControllerConfig{
+					HeartbeatPeriod: 150 * time.Second,
+					RefreshInterval: 15 * time.Minute,
+					AuthInfosPath:   "/var/IEF/secret",
+					PlacementURL:    "https://10.154.193.32:7444/v1/placement_external/message_queue",
```
Comment (rohitsardesai83) [inline pkg/edgehub/controller_test.go]: are these required to be actual url's ? for tests to run , we can have some fake urls right ?

Comment (rohitsardesai83) [reply pkg/edgehub/controller_test.go]: better not to expose ip:port and such uris in tests.

File: pkg/edgehub/controller_test.go [hunk 31]
```diff
@@ -0,0 +1,820 @@
...
+//TestGetCloudHubUrl() tests the procurement of the cloudHub URL from the placement server
+func TestGetCloudHubUrl(t *testing.T) {
+	tests := []struct {
+		name            string
+		fields          Controller
+		webSocketConfig config.WebSocketConfig
+		want            string
+		expectedError   error
+	}{
+		{"TestController_getCloudHubUrl: Proper Request", Controller{
```
Comment (rohitsardesai83) [inline pkg/edgehub/controller_test.go]: is it necessary to have every test case name prefixed with the test case ? won't go test describe the test which failed ?
we could have short names like "get valid cloud hub url". this applies to all tests throughout which have such descriptions.

**Comment by sids-b** [comment]
@sujithsimon22 , please rebase your fork to latest for passing CI checks

File: pkg/edgehub/controller_test.go [hunk 32]
```diff
@@ -0,0 +1,815 @@
...
+			ehc := &Controller{
+				context:    tt.fields.context,
+				chClient:   tt.fields.chClient,
+				config:     tt.fields.config,
+				stopChan:   tt.fields.stopChan,
+				syncKeeper: tt.fields.syncKeeper,
+				keeperLock: tt.fields.keeperLock,
+			}
+			if tt.receiveErr {
+				mockAdapter.EXPECT().Receive().Return(*model.NewMessage(""), errors.New("Connection Refused")).AnyTimes()
```
Comment (sids-b) [inline pkg/edgehub/controller_test.go]: can we parameterize the return and times of mock statements to remove if else here ?
